web/packages/teleport/src/UnifiedResources/FilterPanel.tsx

```diff
@@ -0,0 +1,156 @@
+import React from 'react';
+import Select from 'shared/components/Select';
+import styled from 'styled-components';
+
+import { encodeUrlQueryParams } from 'teleport/components/hooks/useUrlFiltering';
+import { AgentFilter, SortType } from 'teleport/services/agents';
+
+const kindOptions = [
+  { label: 'Application', value: 'app' },
+  { label: 'Database', value: 'db' },
+  { label: 'Desktop', value: 'windows_desktop' },
+  { label: 'Kubernetes', value: 'kube_cluster' },
+  { label: 'Server', value: 'node' },
+];
+
+const sortFieldOptions = [
+  { label: 'Name', value: 'name' },
+  { label: 'Type', value: 'kind' },
+];
+
+export interface FilterPanelProps {
+  pathname: string;
+  replaceHistory: (path: string) => void;
+  params: AgentFilter;
+  setParams: (params: AgentFilter) => void;
+  setSort: (sort: SortType) => void;
+}
+
+export function FilterPanel({
+  pathname,
+  replaceHistory,
+  params,
+  setParams,
+  setSort,
+}: FilterPanelProps) {
+  const { sort, kinds } = params;
+
+  const activeSortFieldOption = sortFieldOptions.find(
+    opt => opt.value === sort.fieldName
+  );
+
+  const activeKindOptions = kindOptions.filter(
+    opt => kinds && kinds.includes(opt.value)
+  );
+
+  const onKindsChanged = (filter: any) => {
+    const kinds = (filter ?? []).map(f => f.value);
+    setParams({ ...params, kinds });
+    // TODO(bl-nero): We really shouldn't have to do it, that's what setParams
+    // should be for.
+    const isAdvancedSearch = !!params.query;
+    replaceHistory(
+      encodeUrlQueryParams(
+        pathname,
+        params.search ?? params.query,
+        params.sort,
+        kinds,
+        isAdvancedSearch
+      )
+    );
+  };
+
+  const onSortFieldChange = (option: any) => {
+    setSort({ ...sort, fieldName: option.value });
+  };
+
+  const onSortOrderButtonClicked = () => {
+    setSort(oppositeSort(sort));
+  };
+
+  return (
+    <Flex justifyContent="space-between" mb={2}>
+      <Box width="300px">
+        <FilterSelect
+          isMulti={true}
+          placeholder="Type"
+          options={kindOptions}
+          value={activeKindOptions}
+          onChange={onKindsChanged}
+        />
+      </Box>
+      <Flex>
+        <Box width="100px">
+          <SortSelect
+            options={sortFieldOptions}
+            value={activeSortFieldOption}
+            isSearchable={false}
+            onChange={onSortFieldChange}
+          />
+        </Box>
+        <SortOrderButton px={3} onClick={onSortOrderButtonClicked}>
+          {sort.dir === 'ASC' && <icons.SortAsc />}
+          {sort.dir === 'DESC' && <icons.SortDesc />}
+        </SortOrderButton>
+      </Flex>
+    </Flex>
+  );
+  return null;
+}
+
+function oppositeSort(sort: SortType): SortType {
+  switch (sort.dir) {
+    case 'ASC':
+      return { ...sort, dir: 'DESC' };
+    case 'DESC':
+      return { ...sort, dir: 'ASC' };
+    default:
+      // Will never happen. Of course.
+      return sort;
+  }
```

Comment:
this is fine. we could also just check for `DESC` and make the default the `ASC` case

Reply:
I somehow prefer to be a bit more explicit :)